packages/caravan-fees/src/btcTransactionTemplate.ts

```diff
@@ -0,0 +1,506 @@
+   */
+  get estimatedVsize(): number {
+    return estimateTransactionVsize({
+      addressType: this._scriptType,
+      numInputs: this._inputs.length,
+      numOutputs: this._outputs.length,
+      m: this._requiredSigners,
+      n: this._totalSigners,
+    });
+  }
+
+  /**
+   * Calculates the estimated fee rate of the transaction.
+   * @returns {string} The estimated fee rate in satoshis per vbyte
+   */
+  get estimatedFeeRate(): string {
+    return this.calculateFee()
+      .dividedBy(this.calculateEstimatedVsize())
+      .toString();
+  }
+
+  /**
+   * Adds an input to the transaction.
+   * @param input - The input to add
+   */
+  addInput(input: BtcTxInputTemplate): void {
+    this._inputs.push(input);
+  }
+
+  /**
+   * Adds an output to the transaction.
+   * @param output - The output to add
+   */
+  addOutput(output: BtcTxOutputTemplate): void {
+    this._outputs.push(output);
+  }
+
+  /**
+   * Removes an output from the transaction.
+   * @param index - The index of the output to remove
+   */
+  removeOutput(index: number): void {
+    this._outputs.splice(index, 1);
+  }
+
+  /**
+   * Adjusts the change output of the transaction.
+   * This method calculates a new change amount based on the current inputs,
+   * non-change outputs, and the target fee. It then updates the change output
+   * or removes it if the new amount is below the dust threshold.
+   *
+   * The method also handles cases where the change output already has an amount:
+   * 1. It calculates the difference between the new and current change amount.
+   * 2. If removing change, it ensures the fee doesn't unexpectedly increase.
+   */
+  adjustChangeOutput(): void {
+    if (this.malleableOutputs.length === 0) return;
+
+    // TO DO (MRIGESH):
+    // To handle for multiple change outputs
+    //
+    const changeOutput = this.malleableOutputs[0];
+    const totalOutWithoutChange = this.calculateTotalOutputAmount().minus(
+      this.calculateChangeAmount(),
+    );
+    const currentChangeAmount = new BigNumber(changeOutput.amountSats);
+
+    const newChangeAmount = this.calculateTotalInputAmount()
+      .minus(totalOutWithoutChange)
+      .minus(new BigNumber(this.targetFeesToPay));
+
+    if (newChangeAmount.gte(0)) {
+      const changeAmountDiff = newChangeAmount.minus(currentChangeAmount);
+
+      if (!changeAmountDiff.isZero()) {
+        changeOutput.addAmount(changeAmountDiff.toString());
+      }
+      // get current out amount after adjustment
+
+      const balanceCheck = this.calculateTotalInputAmount()
+        .minus(this.calculateTotalOutputAmount())
+        .minus(new BigNumber(this.targetFeesToPay));
+
+      if (!balanceCheck.isZero()) {
+        throw new Error(
+          `Transaction does not balance. Discrepancy: ${balanceCheck.toString()} satoshis`,
+        );
+      }
+    } else {
+      throw new Error(
+        `New Change Amount ${newChangeAmount.toString()} cannot be negative`,
+      );
+    }
```

Comment:
this is a nit, but might be cleaner to just do this at the top:
```suggestion
    if (newChangeAmount.lt(0) {
      throw new Error(
        `New Change Amount ${newChangeAmount.toString()} cannot be negative`,
      );
    }
```
And then you don't have to have the main logic nested in an if statement. Basically throw if it's negative otherwise keep going.

Reply:
thanks for the suggestion ... changes [done](https://github.com/caravan-bitcoin/caravan/pull/114/commits/09d9bc5af71e28a76c78abd005d84fdd2d25b2aa#diff-c113a33821d4e86e4ad4f20743227a51337831792cc5c5926aa2a5971c11d7b1R252)